### PR TITLE
Fix view renderer for nested views

### PR DESCRIPTION
**Pull request by ddworak**

### CODE REVIEW

File: core/.js/src/test/scala/io/udash/testing/TestViewFactory.scala [hunk 1]
```diff
@@ -14,31 +15,42 @@ class TestViewFactory[T <: TestState] extends ViewFactory[T] {
...
+  private def content: String = overrideContent.getOrElse(super.toString)
+
   override def renderChild(view: Option[View]): Unit = {
     super.renderChild(view)
     lastChild = view.orNull
   }
 
+  override def clearChildViewContainer(): Unit = {
+    super.clearChildViewContainer()
+    lastChild = null
```
Comment (bgrochal) [inline core/.js/src/test/scala/io/udash/testing/TestViewFactory.scala:35]: Is that necessary? The only way to call `ContainerView#clearChildViewContainer` (except for subclasses) goes through a `ContainerView#renderChild` invocation, which handles `lastChild` properly.

In fact, I'd consider making `ContainerView#clearChildViewContainer` private.

**Review by bgrochal — APPROVED** [review]
Some suggestions left to be considered, but LGTM generally. I like the way it works now - clear and consistent, nicely done 👍

File: core/.js/src/main/scala/io/udash/view/ViewRenderer.scala [hunk 2]
```diff
@@ -53,24 +42,29 @@ private[udash] class ViewRenderer(rootElement: => Element) {
...
-    } else {
-      val removedViews = views.size - currentViewsToLeaveSize
-      views.trimEnd(removedViews)
-      val rootView = views.last
-      val rootViewToAttach = mergeViews(pathToAdd.iterator)
-      if (removedViews > 0 || rootViewToAttach.isDefined) renderChild(rootView, rootViewToAttach)
+    val viewsToLeaveSize = findEqPrefix(subPathToLeave, views.iterator).size
+    val rootView = views.applyOpt(viewsToLeaveSize - 1)
+    //technically e.g. B from docs stays, but we run it through the algorithm anyway for proper children cleanup
+    val unmodifiedViews = rootView.foldLeft(viewsToLeaveSize)((unmodified, _) => unmodified - 1)
```
[the remark this answers is not in the document]
Comment (ddworak) [reply core/.js/src/main/scala/io/udash/view/ViewRenderer.scala:58]: Not a strong preference, but I'd rather not repeat this very weakly typed variable.

File: core/.js/src/test/scala/io/udash/testing/TestViewFactory.scala (hunk 1 above)
Comment (ddworak) [reply core/.js/src/test/scala/io/udash/testing/TestViewFactory.scala:35]: `clearChildViewContainer` cannot be private - it is a hook which may be used by `ContainerView` implementations in the wild and the only clean way to attach user code to container cleanup.

This change is not necessary here, but it's factually correct. In fact `renderChild` does too much if it's supposed to resemble real interaction - it should only do `view.foreach`. I found this out while testing different approaches to handling this issue and I think lying in these tests makes it harder to debug these cases.

File: core/.js/src/main/scala/io/udash/view/ViewRenderer.scala [hunk 3]
```diff
@@ -53,24 +42,29 @@ private[udash] class ViewRenderer(rootElement: => Element) {
...
-    if (currentViewsToLeaveSize == 0) {
-      require(pathToAdd.nonEmpty, "You cannot remove all views, without adding any new view.")
-      replaceCurrentViews(pathToAdd)
-    } else {
-      val removedViews = views.size - currentViewsToLeaveSize
-      views.trimEnd(removedViews)
-      val rootView = views.last
-      val rootViewToAttach = mergeViews(pathToAdd.iterator)
-      if (removedViews > 0 || rootViewToAttach.isDefined) renderChild(rootView, rootViewToAttach)
+    val viewsToLeaveSize = findEqPrefix(subPathToLeave, views.iterator).size
```
[the remark this answers is not in the document]
Comment (ddworak) [reply core/.js/src/main/scala/io/udash/view/ViewRenderer.scala:55]: I agree, I'm just not sure what to do with `"check old views hierarchy and use only valid prefix"` test case, which intentionally provides an incorrect subpath. I cannot yet think of a case, where `RoutingEngine` would do that. Being bulletproof in isolation may not be such a bad thing anyway for `ViewRenderer`.

**Pull request merged**